cosima_cookbook/database.py

```diff
@@ -600,16 +568,60 @@ def build_index(
         directories = [directories]
 
     indexed = 0
-    for directory in directories:
-        indexed += index_experiment(
-            directory, session, client, update, prune, delete, followsymlinks
-        )
+    for directory in [Path(d) for d in directories]:
+        # find all netCDF files in the experiment directory
+        files = find_files(directory, followsymlinks=followsymlinks)
+        expt = find_experiment(session, directory)
+
+        if expt is None:
+            expt = NCExperiment(
+                experiment=str(directory.name), root_dir=str(directory.resolve())
+            )
+        else:
+            if not update:
+                print(
+                    "Not re-indexing experiment: {}\n"
+                    "Pass `update=True` to build_index()".format(directory.name)
+                )
+                continue
+
+        print("Indexing experiment: {}".format(directory.name))
+
+        if prune:
+            # Prune files that exist in the database but are not present
+            # on the filesystem
+            missing_files = set([f.ncfile for f in expt.ncfiles]) - files
+            if len(missing_files) > 0:
+                # Make a list of NCFile objects to pass to prune_files
+                missing_files_objs = [
+                    f for f in expt.ncfiles if f.ncfile in missing_files
+                ]
+                prune_files(expt, session, missing_files_objs, delete=delete)
+
+        # Filter conditions
+        if not force:
+            # Only pass files that are not already in DB
+            files = files - set([f.ncfile for f in expt.ncfiles])
+
+        indexed += index_experiment(files, session, expt, client)
+
+        # if everything went smoothly, commit these changes to the database
+        session.commit()
```

Comment:
I've moved the `session.commit()` inside the loop over the `directories` to make it more robust when encountering errors, so subsequent calls to `build_index` won't have to re-index stuff that worked ok last time.

Reply:
I think that's reasonable, we've already made sure the experiments themselves are consistent (uniqueness enforced, etc.) before this point.